sled-agent/src/illumos/dladm.rs

```diff
@@ -75,11 +82,77 @@ pub struct DeleteVnicError {
 #[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
 pub struct PhysicalLink(pub String);
 
+/// The name of an etherstub
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct Etherstub(pub String);
+
+/// The name of an etherstub's underlay VNIC
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct EtherstubVnic(pub String);
+
+/// Identifies that an object may be used to create a VNIC.
+pub trait VnicSource {
+    fn name(&self) -> &str;
+}
+
+impl VnicSource for Etherstub {
+    fn name(&self) -> &str {
+        &self.0
+    }
+}
+
+impl VnicSource for PhysicalLink {
+    fn name(&self) -> &str {
+        &self.0
+    }
+}
+
 /// Wraps commands for interacting with data links.
 pub struct Dladm {}
 
 #[cfg_attr(test, mockall::automock, allow(dead_code))]
 impl Dladm {
+    /// Creates an etherstub, or returns one which already exists.
+    pub fn create_etherstub() -> Result<Etherstub, ExecutionError> {
+        if let Ok(stub) = Self::get_etherstub() {
+            return Ok(stub);
+        }
+        let mut command = std::process::Command::new(PFEXEC);
+        let cmd =
+            command.args(&[DLADM, "create-etherstub", "-t", ETHERSTUB_NAME]);
```

Comment:
Just curious, why create the etherstub as temporary? Does this code run at each boot?

Reply:
It's somewhat arbitrary, but it basically matches how we manage a lot of resources on the Sled (such as VNICs, etc): the sled agent always recreates them on boot.

We definitely *could* let it be persistent, but the transient nature makes the "boot on a new machine" vs "reboot an old machine" cases seem a lot more similar, which is handy - especially when the cost of creating these devices is relatively low.